Provide Docker target and discovery in Promtail. (#4911)
grafana/loki: docs/sources/clients/promtail/configuration.md

# running on the same host as Promtail.
consulagent_sd_configs:
  [ - <consulagent_sd_config> ... ]

> # Describes how to use the Docker daemon API to discover containers running on the same host as Promtail.

Reviewer: Inject a newline, so this line doesn't cause the need for a horizontal scroll bar.

```suggestion
# Describes how to use the Docker daemon API to discover containers running on
# the same host as Promtail.
```

You can leverage [pipeline stages](pipeline_stages) if, for example, you want to parse the JSON log line and extract more labels or change the log line format.

### Docker

The `docker` block configures Promtail to pull logs for a Docker container via the Docker daemon. 

```yaml
# Address of the Docker daemon. It defaults to unix:///var/run/docker.sock.
host: <string>

> # The name or id of the container that should be scraped.

Author: There are two new configs: `docker` and `docker_sd`. The former targets one specific container. I needed to add the target anyways so I thought I could also add `docker`. If that's too confusing I'll remove it.

You can leverage [pipeline stages](pipeline_stages) if, for example, you want to parse the JSON log line and extract more labels or change the log line format.

### Docker

The `docker` block configures Promtail to pull logs for a Docker container via the Docker daemon. 

```yaml
# Address of the Docker daemon. It defaults to unix:///var/run/docker.sock.
host: <string>

# Optional port of the Docker daemon address.

> post: <int>

Reviewer 2: Guess this should be `port`.

directly which has basic support for filtering nodes (currently by node
metadata and a single tag).

### docker_sd_config

> Docker service discovery configurations allow retrieving targets from a Docker daemon.
> It will only watch containers of the daemon referenced with the host parameter. That
> means it should run on each node in a distributed setup. The containers must run with
> either the [json-file](https://docs.docker.com/config/containers/logging/json-file/)
> or [journald](https://docs.docker.com/config/containers/logging/journald/) logging driver.

Reviewer: Trying here to suggest wording that clarifies "it" in more than one place. If my clarifications are wrong, please revise to clarify correctly.

```suggestion
Docker service discovery allows retrieving targets from a Docker daemon.
It will only watch containers of the Docker daemon referenced with the host parameter. Docker
service discovery should run on each node in a distributed setup. The containers must run with
either the [json-file](https://docs.docker.com/config/containers/logging/json-file/)
or [journald](https://docs.docker.com/config/containers/logging/journald/) logging driver.
```

Please note that the discovery will not pick up finished containers. That means
Promtail will not scrape the remaining logs from finished containers after a restart.

The configuration is inherited from [Prometheus' Docker service discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#docker_sd_config).

```yaml
# Address of the Docker daemon.  Use unix:///var/run/docker.sock for a local setup.
host: <string>

# Optional proxy URL.
[ proxy_url: <string> ]

# TLS configuration.
tls_config:
  [ <tls_config> ]

# The port to scrape metrics from, when `role` is nodes, and for discovered
# tasks and services that don't have published ports.
[ port: <int> | default = 80 ]

# The host to use if the container is in host networking mode.
[ host_networking_host: <string> | default = "localhost" ]

# Optional filters to limit the discovery process to a subset of available
# resources.

> # The available filters are listed in the upstream documentation:

Reviewer: Suggestion made because "upstream" and "downstream" are not well defined. 
```suggestion
# The available filters are listed in the Docker documentation:
```

# Containers: https://docs.docker.com/engine/api/v1.41/#operation/ContainerList
[ filters:
  [ - name: <string>
      values: <string>, [...] ]
]

# The time after which the containers are refreshed.
[ refresh_interval: <duration> | default = 60s ]

> # Authentication information used to authenticate to the Docker daemon.

Reviewer: Can you please clarify what is being authenticated to the Docker daemon?

Author: I've tried to clarify that Promtail is authenticated.

# Note that `basic_auth` and `authorization` options are
# mutually exclusive.
# password and password_file are mutually exclusive.

# Optional HTTP basic authentication information.
basic_auth:
  [ username: <string> ]
  [ password: <secret> ]
  [ password_file: <string> ]

# Optional `Authorization` header configuration.
authorization:
  # Sets the authentication type.
  [ type: <string> | default: Bearer ]
  # Sets the credentials. It is mutually exclusive with
  # `credentials_file`.
  [ credentials: <secret> ]
  # Sets the credentials to the credentials read from the configured file.
  # It is mutually exclusive with `credentials`.
  [ credentials_file: <filename> ]

# Optional OAuth 2.0 configuration.
# Cannot be used at the same time as basic_auth or authorization.
oauth2:
  [ <oauth2> ]

# Configure whether HTTP requests follow HTTP 3xx redirects.
[ follow_redirects: <bool> | default = true ]
```

Available meta labels:

>   * `__meta_docker_container_id`: the id of the container

Reviewer: ```suggestion
  * `__meta_docker_container_id`: the ID of the container
```

  * `__meta_docker_container_name`: the name of the container
  * `__meta_docker_container_network_mode`: the network mode of the container
  * `__meta_docker_container_label_<labelname>`: each label of the container
  * `__meta_docker_container_log_stream`: the log stream type `stdout` or `stderr`
  * `__meta_docker_network_id`: the ID of the network
  * `__meta_docker_network_name`: the name of the network
  * `__meta_docker_network_ingress`: whether the network is ingress
  * `__meta_docker_network_internal`: whether the network is internal
  * `__meta_docker_network_label_<labelname>`: each label of the network
  * `__meta_docker_network_scope`: the scope of the network
  * `__meta_docker_network_ip`: the IP of the container in this network
  * `__meta_docker_port_private`: the port on the container
  * `__meta_docker_port_public`: the external port if a port-mapping exists
  * `__meta_docker_port_public_ip`: the public IP if a port-mapping exists

> These labels can be used during relabelling. For instance, the following configuration scrapes the container named `flog` and removes the leading `/` form the container name.

Reviewer: ```suggestion
These labels can be used during relabeling. For instance, the following configuration scrapes the container named `flog` and removes the leading slash (`/`) from the container name.
```

```yaml
scrape_configs:
  - job_name: flog_scrape 
    docker_sd_configs:
      - host: unix:///var/run/docker.sock
        refresh_interval: 5s
        filters:
          - name: name
            values: [flog] 
    relabel_configs:
      - source_labels: ['__meta_docker_container_name']

>         regex: '/(.*)'

Reviewer 3: Not sure you need regex I think you can use keep action

Author: The name is `/flog`. How would I strip the leading `/`? It's just an example.

[ filters:
  [ - name: <string>
      values: <string>, [...] ]
]

# The time after which the containers are refreshed.
[ refresh_interval: <duration> | default = 60s ]

# Authentication information used to authenticate to the Docker daemon.
# Note that `basic_auth` and `authorization` options are
# mutually exclusive.
# password and password_file are mutually exclusive.

# Optional HTTP basic authentication information.
basic_auth:
  [ username: <string> ]
  [ password: <secret> ]
  [ password_file: <string> ]

# Optional `Authorization` header configuration.
authorization:
  # Sets the authentication type.
  [ type: <string> | default: Bearer ]
  # Sets the credentials. It is mutually exclusive with
  # `credentials_file`.
  [ credentials: <secret> ]
  # Sets the credentials to the credentials read from the configured file.
  # It is mutually exclusive with `credentials`.
  [ credentials_file: <filename> ]

# Optional OAuth 2.0 configuration.
# Cannot be used at the same time as basic_auth or authorization.
oauth2:
  [ <oauth2> ]

# Configure whether HTTP requests follow HTTP 3xx redirects.
[ follow_redirects: <bool> | default = true ]
```

Available meta labels:

>   * `__meta_docker_container_id`: the id of the container

Reviewer 3: Would be great to have a full good example, with correct relabeling so people can copy/paste.

directly which has basic support for filtering nodes (currently by node
metadata and a single tag).

### docker_sd_config

Docker service discovery configurations allow retrieving targets from a Docker daemon.
It will only watch containers of the daemon referenced with the host parameter. That
means it should run on each node in a distributed setup. The containers must run with
either the [json-file](https://docs.docker.com/config/containers/logging/json-file/)
or [journald](https://docs.docker.com/config/containers/logging/journald/) logging driver.

Please note that the discovery will not pick up finished containers. That means
Promtail will not scrape the remaining logs from finished containers after a restart.

> The configuration is inherited from [Prometheus' Docker service dicsovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#docker_sd_config).

Reviewer 2: ```suggestion
The configuration is inherited from [Prometheus' Docker service discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#docker_sd_config).
```

```yaml
# Address of the Docker daemon.  Use unix:///var/run/docker.sock for a local setup.
host: <string>

# Optional proxy URL.
[ proxy_url: <string> ]

# TLS configuration.
tls_config:
  [ <tls_config> ]

# The port to scrape metrics from, when `role` is nodes, and for discovered
# tasks and services that don't have published ports.
[ port: <int> | default = 80 ]

# The host to use if the container is in host networking mode.
[ host_networking_host: <string> | default = "localhost" ]

# Optional filters to limit the discovery process to a subset of available
# resources.
# The available filters are listed in the upstream documentation:
# Containers: https://docs.docker.com/engine/api/v1.41/#operation/ContainerList

> [ filters:
>   [ - name: <string>
>       values: <string>, [...] ]

Reviewer 2: Missing closing square bracket:

```suggestion
[ filters:
  [ - name: <string>
      values: <string>, [...] ]
]
```

or remove the outer one:

```suggestion
filters:
  [ - name: <string>
      values: <string>, [...] ]
```

# The time after which the containers are refreshed.
[ refresh_interval: <duration> | default = 60s ]

# Authentication information used to authenticate to the Docker daemon.
# Note that `basic_auth` and `authorization` options are
# mutually exclusive.
# password and password_file are mutually exclusive.

# Optional HTTP basic authentication information.
basic_auth:
  [ username: <string> ]
  [ password: <secret> ]
  [ password_file: <string> ]

# Optional `Authorization` header configuration.
authorization:
  # Sets the authentication type.
  [ type: <string> | default: Bearer ]
  # Sets the credentials. It is mutually exclusive with
  # `credentials_file`.
  [ credentials: <secret> ]
  # Sets the credentials to the credentials read from the configured file.
  # It is mutually exclusive with `credentials`.
  [ credentials_file: <filename> ]

# Optional OAuth 2.0 configuration.
# Cannot be used at the same time as basic_auth or authorization.
oauth2:
  [ <oauth2> ]

# Configure whether HTTP requests follow HTTP 3xx redirects.
[ follow_redirects: <bool> | default = true ]
```

Available meta labels:

>   * `_meta_docker_container_id`: the id of the container

Reviewer 2: ```suggestion
  * `__meta_docker_container_id`: the id of the container
```
